File: content-solutions/backtracking/flower-sudoku.py
def solution(garden):
    grid = [list(row) for row in garden]
    blanks = [(r, c) for r in range(6) for c in range(6) if grid[r][c] == "."]

    def fits(r, c, flower):
        for i in range(6):
            if grid[r][i] == flower or grid[i][c] == flower:
                return False
        top, left = r // 2 * 2, c // 3 * 3
        for i in range(top, top + 2):
            for j in range(left, left + 3):
                if grid[i][j] == flower:
                    return False
        return True

    def fill(k):
        if k == len(blanks):
            return True
        r, c = blanks[k]
        for flower in "123456":
            if fits(r, c, flower):
                grid[r][c] = flower
                if fill(k + 1):
                    return True
                grid[r][c] = "."
        return False

    fill(0)
    return ["".join(row) for row in grid]

File: content-solutions/backtracking/flower-sudoku.js
function solution(garden) {
  const grid = garden.map((row) => row.split(""));
  const blanks = [];
  for (let r = 0; r < 6; r++) for (let c = 0; c < 6; c++) if (grid[r][c] === ".") blanks.push([r, c]);
  function fits(r, c, flower) {
    for (let i = 0; i < 6; i++) {
      if (grid[r][i] === flower || grid[i][c] === flower) return false;
    }
    const top = Math.floor(r / 2) * 2;
    const left = Math.floor(c / 3) * 3;
    for (let i = top; i < top + 2; i++) {
      for (let j = left; j < left + 3; j++) if (grid[i][j] === flower) return false;
    }
    return true;
  }
  function fill(k) {
    if (k === blanks.length) return true;
    const [r, c] = blanks[k];
    for (const flower of "123456") {
      if (fits(r, c, flower)) {
        grid[r][c] = flower;
        if (fill(k + 1)) return true;
        grid[r][c] = ".";
      }
    }
    return false;
  }
  fill(0);
  return grid.map((row) => row.join(""));
}
